### DISCUSSION TITLE
Hardware recommendations to implement CAN to WiFi interface?

### DISCUSSION THREAD

**Discussion by blmorris** [micropython/micropython#16301] [Hardware & Peripherals] [answered]
Hi All,
I'm looking for options to implement an interface to make health and status information from a power supply available to a local WiFi network. The power supply provides the information via a CAN interface, so one CAN port is a requirement. The entire system is going to be inside of an industrial steel (NEMA) enclosure, so a connector for an external WiFi antenna is also a hard requirement- any board with only an onboard antenna won't work for this application.
Beyond that, my application would only need a few GPIO for basic board control and indicator LEDs. 

I would be happy to use the Pyboard D-Series - they meet my requirements and I'm somewhat familiar developing with micropython on STM32, but unfortunately they appear to be out of stock: https://store.micropython.org/category/pyboard%20D-series

So far, all of the ESP32 or RP2040 based boards that I've seen have on-board chip or trace antennas; I probably haven't done an exhaustive search.

I've been out of the loop on micropython development for a while, so I'm not really up to date on supported hardware, but I can see that there are a lot more options now than I remember, it is fantastic to see all of the progress!

If there is a resource to help select from supported boards, I'd be happy to use that rather than ask the community to hunt for me. Otherwise I'll be digging through the list of supported boards to determine availability and whether it has the external connector.

Thanks, Bryan

**Comment by davefes** [comment, marked as answer]
ESP32 WROOM 32U for external antennas.  Just checked if CAN is available:
https://github.com/orgs/micropython/discussions/10535
https://github.com/micropython/micropython/pull/9532
I don't know the current status.

**Reply by blmorris** [reply]
Thanks, that is helpful!
Looking for that, I now see that Adafruit does have several ESP32 boards with u.FL connectors for external antennas, so that is a good start.
Going to dig into status of CAN support in ESP32 now.

**Reply by blmorris** [reply]
> I don't know the current status.

So, after chasing a few threads to get up to date, current status seems to be that there are plenty of people with interest in getting a consistent CAN API across all of the ports, but no consensus for exactly how to do that across all the different hardware implementations.
Some relevant discussion threads (#12337) and PRs (#12324, #12330, #12331, #13149, #15989, probably others) and plenty of active discussion, but it doesn't look like we are on the verge of having a either a cross-platform or esp32-specific CAN module in the main trunk.

Will keep my eye on this space and see if I can help out.
